File: LeetCode/Python/Medium/340-Longest-Substring-With-At-Most-K-Distinct-Characters/solution.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        result: int=0

        if k==0:
            return result

        n: int=len(s)

        hashmap: list[int]=[0] * 128

        left: int=0

        right: int=0

        distinct: int=0

        while right < n:
            right_letter: chr=s[right]

            right_ascii: int=ord(right_letter)

            #If this is the first occurrence of the chracter
            if hashmap[right_ascii]==0:
                #Assign initial frequency
                hashmap[right_ascii]=1

                #Increment the number of distinct characters
                distinct+=1
            else:
                #Just increment the frequency
                hashmap[right_ascii]+=1

            #At this point, if the current window has more than k distinct characters, then we need to shrink the window from the left
            while left < right and distinct > k:
                left_letter: chr=s[left]

                left_ascii: int=ord(left_letter)

                #If the frequency of s[left] is greater than zero
                if hashmap[left_ascii] > 0:
                    hashmap[left_ascii]-=1

                    #If the frequency is zero, then this character is no longer in the current window and the number of distinct characters decreases by one
                    if hashmap[left_ascii]==0:
                        distinct-=1

                #Move the left pointer
                left+=1

            #At this point, the current window has at most k distinct characters so calculate its length
            length: int=(right - left) + 1

            #Update the result
            result=max(result, length)

            #Move the right pointer
            right+=1

        return result

File: LeetCode/Rust/Medium/340-Longest-Substring-with-At-Most-K-Distinct-Characters/solution.rs
use std::collections::{HashMap};

impl Solution
{
    pub fn length_of_longest_substring_k_distinct(s: String, k: i32) -> i32
    {
        let mut result: i32=0;

        if k==0
        {
            return result;
        }

        let n: usize=s.len();

        let mut hashmap: HashMap<char, i32>=HashMap::new();

        let characters: Vec<char>=s.chars().collect::<Vec<char>>();

        let mut left: usize=0;

        let mut right: usize=0;

        while right < n
        {
            let right_character: char=characters[right];

            hashmap.entry(right_character).and_modify(|frequency| *frequency+=1).or_insert(1);

            while (hashmap.len() as i32) > k
            {
                let left_character: char=characters[left];

                hashmap.entry(left_character).and_modify(|frequency| *frequency-=1);

                if *hashmap.get(&left_character).unwrap()==0
                {
                    hashmap.remove(&left_character);
                }

                left+=1;
            }

            let length: usize=(right - left) + 1;

            result=std::cmp::max(result, length as i32);

            right+=1;
        }

        result
    }
}